contracts/src/token/erc1155/extensions/supply.rs

```diff
@@ -0,0 +1,881 @@
+    /// then the error [`erc1155::Error::InsufficientBalance`] is returned.
+    /// If the `from` is not the caller (`msg::sender()`),
+    /// and the caller does not have the right to approve, then the error
+    /// [`erc1155::Error::MissingApprovalForAll`] is returned.
+    /// If [`erc1155::IERC1155Receiver::on_erc_1155_batch_received`] hasn't
+    /// returned its interface id or returned with error, then the error
+    /// [`erc1155::Error::InvalidReceiver`] is returned.
+    ///
+    /// # Requirements
+    ///
+    /// * `to` cannot be the `Address::ZERO`.
+    /// * If the caller is not `from`, it must have been approved to spend
+    ///   `from`'s tokens via [`IErc1155::set_approval_for_all`].
+    /// * `from` must have a balance of tokens being transferred of at least
+    ///   transferred amount.
+    /// * `ids` and `values` must have the same length.
+    /// * If `to` refers to a smart contract, it must implement
+    ///   [`erc1155::IERC1155Receiver::on_erc_1155_batch_received`] and return
+    ///   the acceptance magic value.
+    ///
+    /// # Events
+    ///
+    /// Emits either a [`TransferSingle`] or a [`erc1155::TransferBatch`] event,
+    /// depending on the length of the array arguments.
+    ///
+    /// # Panics
+    ///
+    /// Should not panic.
+    pub fn safe_batch_transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        ids: Vec<U256>,
+        values: Vec<U256>,
+        data: Bytes,
+    ) -> Result<(), erc1155::Error> {
+        self.erc1155.authorize_transfer(from)?;
+        self.do_safe_transfer_from(from, to, ids, values, &data)
+    }
+}
+
+impl Erc1155Supply {
+    // Note: overriding `_update` requires reimplementing all of the functions
+    // that use it.
+
+    /// Override of [`Erc1155::_update`] that updates the supply of tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `from` - Account of the sender.
+    /// * `to` - Account of the recipient.
+    /// * `token_ids` - Array of all token id.
+    /// * `values` - Array of all amount of tokens to be supplied.
+    ///
+    /// # Errors
+    ///
+    /// If length of `ids` is not equal to length of `values`, then the
+    /// error [`erc1155::Error::InvalidArrayLength`] is returned.
+    /// If `value` is greater than the balance of the `from` account,
+    /// then the error [`erc1155::Error::InsufficientBalance`] is returned.
+    ///
+    /// NOTE: The ERC-1155 acceptance check is not performed in this function.
+    /// See [`Self::_update_with_acceptance_check`] instead.
+    ///
+    /// # Events
+    ///
+    /// Emits a [`erc1155::TransferSingle`] event if the arrays contain one
+    /// element, and [`erc1155::TransferBatch`] otherwise.
+    ///
+    /// # Panics
+    ///
+    /// If updated balance and/or supply exceeds `U256::MAX`, may happen during
+    /// the `mint` operation.
+    pub fn _update(
```

Comment:
I see that `_update` and `_update_with_acceptance_check` should not be marked `pub`. And probably some other functions..